File: web/src/protocol.ts
export const PROTOCOL_VERSION = 1;

export type ClientMessage =
  | {
      type: 'open';
      protocol: number;
      cols: number;
      rows: number;
      profile: string;
    }
  | { type: 'resize'; cols: number; rows: number }
  | { type: 'close' };

export type ServerMessage =
  | {
      type: 'opened';
      protocol: number;
      sessionId: string;
      cols: number;
      rows: number;
    }
  | { type: 'exit'; code: number; signal: string | null }
  | { type: 'error'; code: string; message: string; fatal: boolean };

export function encodeControl(message: ClientMessage): string {
  return JSON.stringify(message);
}

export function parseServerMessage(data: string): ServerMessage | null {
  try {
    const message: unknown = JSON.parse(data);
    if (!isRecord(message) || typeof message.type !== 'string') {
      return null;
    }
    if (message.type === 'opened') {
      return isNumber(message.protocol) &&
        typeof message.sessionId === 'string' &&
        isNumber(message.cols) &&
        isNumber(message.rows)
        ? {
            type: 'opened',
            protocol: message.protocol,
            sessionId: message.sessionId,
            cols: message.cols,
            rows: message.rows,
          }
        : null;
    }
    if (message.type === 'exit') {
      return isNumber(message.code) &&
        (message.signal === null || typeof message.signal === 'string')
        ? { type: 'exit', code: message.code, signal: message.signal }
        : null;
    }
    if (message.type === 'error') {
      return typeof message.code === 'string' &&
        typeof message.message === 'string' &&
        typeof message.fatal === 'boolean'
        ? {
            type: 'error',
            code: message.code,
            message: message.message,
            fatal: message.fatal,
          }
        : null;
    }
    return null;
  } catch {
    return null;
  }
}

export function encodeTerminalInput(data: string): Uint8Array {
  return new TextEncoder().encode(data);
}

export function encodeBinaryInput(data: string): Uint8Array {
  const bytes = new Uint8Array(data.length);
  for (let index = 0; index < data.length; index += 1) {
    bytes[index] = data.charCodeAt(index) & 0xff;
  }
  return bytes;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}

function isNumber(value: unknown): value is number {
  return typeof value === 'number' && Number.isFinite(value);
}

File: crates/rssh-web/src/protocol.rs
use serde::{Deserialize, Serialize};

pub const PROTOCOL_VERSION: u16 = 1;
pub const MIN_COLUMNS: u16 = 2;
pub const MAX_COLUMNS: u16 = 500;
pub const MIN_ROWS: u16 = 1;
pub const MAX_ROWS: u16 = 300;
pub const MAX_INPUT_FRAME_BYTES: usize = 64 * 1024;
pub const MAX_CONTROL_FRAME_BYTES: usize = 8 * 1024;

#[derive(Debug, Deserialize, PartialEq, Eq)]
#[serde(tag = "type")]
pub enum ClientMessage {
    #[serde(rename = "open")]
    Open {
        protocol: u16,
        cols: u16,
        rows: u16,
        profile: String,
    },
    #[serde(rename = "resize")]
    Resize { cols: u16, rows: u16 },
    #[serde(rename = "close")]
    Close,
}

#[derive(Debug, Serialize, PartialEq, Eq)]
#[serde(tag = "type")]
pub enum ServerMessage<'a> {
    #[serde(rename = "opened")]
    Opened {
        protocol: u16,
        #[serde(rename = "sessionId")]
        session_id: &'a str,
        cols: u16,
        rows: u16,
    },
    #[serde(rename = "exit")]
    Exit { code: u32, signal: Option<&'a str> },
    #[serde(rename = "error")]
    Error {
        code: &'a str,
        message: &'a str,
        fatal: bool,
    },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TerminalDimensions {
    pub cols: u16,
    pub rows: u16,
}

impl TerminalDimensions {
    /// Validates a browser-provided terminal size against the protocol limits.
    ///
    /// # Errors
    ///
    /// Returns which dimension is outside the supported range.
    pub fn validate(cols: u16, rows: u16) -> Result<Self, DimensionError> {
        if !(MIN_COLUMNS..=MAX_COLUMNS).contains(&cols) {
            return Err(DimensionError::Columns(cols));
        }
        if !(MIN_ROWS..=MAX_ROWS).contains(&rows) {
            return Err(DimensionError::Rows(rows));
        }
        Ok(Self { cols, rows })
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DimensionError {
    Columns(u16),
    Rows(u16),
}

impl DimensionError {
    #[must_use]
    pub const fn code(self) -> &'static str {
        match self {
            Self::Columns(_) => "INVALID_COLUMNS",
            Self::Rows(_) => "INVALID_ROWS",
        }
    }
}

#[cfg(test)]
mod tests {
    use super::{ClientMessage, DimensionError, ServerMessage, TerminalDimensions};

    #[test]
    fn parses_open_message() {
        let message: ClientMessage = serde_json::from_str(
            r#"{"type":"open","protocol":1,"cols":120,"rows":32,"profile":"local-default"}"#,
        )
        .unwrap();
        assert_eq!(
            message,
            ClientMessage::Open {
                protocol: 1,
                cols: 120,
                rows: 32,
                profile: "local-default".to_owned(),
            }
        );
    }

    #[test]
    fn rejects_out_of_range_dimensions() {
        assert_eq!(
            TerminalDimensions::validate(1, 24),
            Err(DimensionError::Columns(1))
        );
        assert_eq!(
            TerminalDimensions::validate(80, 301),
            Err(DimensionError::Rows(301))
        );
    }

    #[test]
    fn serializes_opened_message_with_wire_names() {
        let message = ServerMessage::Opened {
            protocol: 1,
            session_id: "session-1",
            cols: 80,
            rows: 24,
        };
        assert_eq!(
            serde_json::to_string(&message).unwrap(),
            r#"{"type":"opened","protocol":1,"sessionId":"session-1","cols":80,"rows":24}"#
        );
    }
}
